skimage/filters/_multiotsu.pyx

```diff
@@ -6,101 +6,295 @@ import numpy as np
 
-def _find_threshold_multiotsu(double [:, ::1] var_btwcls,
-                              Py_ssize_t classes,
-                              Py_ssize_t bins):
-    """
-    Cython utility function for finding thresholds in multi-Otsu algorithm.
-    This function is only called by filters.threshold_multiotsu.
+
+def _get_multiotsu_thresh_indices_lut(float [::1] prob,
+                                      Py_ssize_t thresh_count):
+
+    cdef Py_ssize_t nbins = prob.shape[0]
+    py_thresh_indices = np.empty(thresh_count, dtype=np.intp)
+    cdef Py_ssize_t[::1] thresh_indices = py_thresh_indices
+    cdef Py_ssize_t[::1] current_indices = np.empty(thresh_count, dtype=np.intp)
+    cdef float [::1] H = np.zeros((nbins * (nbins + 1)) / 2, dtype=np.float32)
+    cdef float [::1] P = np.empty(nbins, dtype=np.float32)
+    cdef float [::1] S = np.empty(nbins, dtype=np.float32)
+
+    with nogil:
+        _set_var_btwcls_lut(prob, nbins, H, P, S)
+        _set_thresh_indices_lut(H, hist_idx=0,
+                                thresh_idx=0, nbins=nbins,
+                                thresh_count=thresh_count, sigma_max=0,
+                                current_indices=current_indices,
+                                thresh_indices=thresh_indices)
+
+    return py_thresh_indices
+
+
+def _get_multiotsu_thresh_indices(float [::1] prob, Py_ssize_t thresh_count):
+
+    cdef Py_ssize_t nbins = prob.shape[0]
+    py_thresh_indices = np.empty(thresh_count, dtype=np.intp)
+    cdef Py_ssize_t[::1] thresh_indices = py_thresh_indices
+    cdef Py_ssize_t[::1] current_indices = np.empty(thresh_count, dtype=np.intp)
+    cdef float [:, ::1] H = np.zeros((nbins, nbins))
+    cdef float [::1] P = np.empty(nbins)
+    cdef float [::1] S = np.empty(nbins)
+
+    with nogil:
+        _set_moment_lut_first_row(prob, nbins, P, S)
+        _set_thresh_indices(P, S, hist_idx=0,
+                            thresh_idx=0, nbins=nbins,
+                            thresh_count=thresh_count, sigma_max=0,
+                            current_indices=current_indices,
+                            thresh_indices=thresh_indices)
+
+    return py_thresh_indices
+
+
+cdef void _set_moment_lut_first_row(float [::1] prob,
+                                    Py_ssize_t nbins,
+                                    float [::1] P,
+                                    float [::1] S) nogil:
+    cdef cnp.intp_t i
+
+    P[0] = prob[0]
+    S[0] = prob[0]
+    for i in range(1, nbins):
+        P[i] = P[i-1] + prob[i]
+        S[i] = S[i-1] + i*prob[i]
+
+
+cdef float _get_var_btwclas(float [::1] P, float [::1] S,
+                            Py_ssize_t i, Py_ssize_t j) nogil:
+
+    cdef float Pij, Sij
+
+    if i == 0:
+        if P[i] > 0:
+            return (S[j]**2)/P[j]
+    else:
+        Pij = P[j] - P[i-1]
+        if Pij > 0:
+            Sij = S[j] - S[i-1]
+            return (Sij**2)/Pij
+    return 0
+
+
+cdef float _set_thresh_indices(float[::1] P, float[::1] S,
```

Comment:
I would rather see a longer (more descriptive) name then single capital letter

Reply:
I kept the notations of the reference paper while developing, but you are right, we can modify those. What do you think of `first_moment` and `second_moment`?

Reply:
I understand the importance of better names for our variables, but in this case, I am kindly against it if we don't refer to the "original" names (variables that appear in the paper) somewhere. The paper does not make it clear to understand what/where the "moments" are, citing them only once as the moments (if I recall correctly). All arguments receive P, S and H in the algorithms.
Therefore, @rfezzani, could you please add a note on the pertinent functions? Something like:
```
zeroth_moment, first_moment, and var_btwcls are referred to as P, S, and H in [1]_.
```
Thanks!